Remake of Sidebar

DCoder1997 opened the pull request:
I changed fully the sidebar to a more beautiful new sidebar with his engine.

With this changes now is possible to make an union of a differents toolbars, for example:
If you have a items for all employees and only a element for personal use, the result is a sidebar with 3 elements.
These 3 elements dependes of you can view across of a role settings.

Changed the ubications enum, from Top, Side, Main to Top, Middle and Bottom. 
The union of the sidebar elements, is ordered by Top > Middle > Bottom and for each element is **ordered** by priority (All toolbars, which you can see according to the role settings, are considered in the sidebar).

The sidebar have 3 states, expanded, only icons and hide and it can be changed using the toggle button located in the top on the left side. These states are saved in localstorage of chrome (this code is allocated in Layout)

When the state of sidebard changes the icon changes across of the state, you can configure a 2 types of icon, icon expanded and not expanded. But if you only have a icon, you can ignore the iconMini property.

This sidebar responds at changes of screen, and it converts to a full screen sidebar if is a device with show and not show states.

The login dropdown now includes the culture settings only if you are logged. If you are not logged the culture is separated.

The version information is removed from the example layout.

**IMPORTANT - These changes require a Layout change to this example:**

```
export default function Layout() {
  const appName = "APPNAME";
  const appIcon = AppContext.toAbsoluteUrl("~/logo.svg");

  const itemStorageKeyExtend = "SIDEBAR_EXPANDS";
  const itemStorageKeyHide = "SIDEBAR_HIDE";

  const [refreshId, setRefreshId] = React.useState(0);
  const [sideMenuVisible] = React.useState(true);
  const [expandedSideMenu, setExpandedSideMenu] = React.useState(window.innerWidth > 991 && window.localStorage.getItem(itemStorageKeyExtend) == "1");
  const [sideMenuHide, setSideMenuHide] = React.useState(window.localStorage.getItem(itemStorageKeyHide) == "1");
  const [fullSidebarView, setFullSidebarView] = React.useState(window.innerWidth < 991);
  const sideMenuOpenRef = useUpdatedRef(expandedSideMenu);

  function handleOpenSideBar(to: boolean) {
    setExpandedSideMenu(to);
    setSideMenuHide(false);
  }

  window.addEventListener("resize", (ev) => {
    if (window.innerWidth <= 991) {
      if (!fullSidebarView) {
        setFullSidebarView(true);
        setExpandedSideMenu(false);
        setSideMenuHide(false);
      }
    }
    else {
      if (fullSidebarView) {
        setFullSidebarView(false);
        setExpandedSideMenu(true);
        setSideMenuHide(false);
      }
    }
  });

  React.useEffect(() => {
    AppContext.Expander.onGetExpanded = () => !sideMenuOpenRef.current;
    AppContext.Expander.onSetExpanded = (isExpanded: boolean) => setExpandedSideMenu(!isExpanded);
  }, []); //Sidebar

  function resetUI() {
    setRefreshId(rID => rID + 1);
  };

  React.useEffect(() => {
    AppContext.setResetUI(resetUI);
    return () => AppContext.setResetUI(() => { });
  }, []);

  React.useEffect(() => {
    window.localStorage.setItem(itemStorageKeyExtend, expandedSideMenu ? "1" : "0");
    window.localStorage.setItem(itemStorageKeyHide, sideMenuHide ? "1" : "0");
  }, [sideMenuHide, expandedSideMenu]);

  return (
    <ErrorBoundary >
      <div id="main" key={refreshId}>
        <Notify />
        <div id="main-container">
          <SidebarContainer
            sidebarVisible={AuthClient.currentUser() && sideMenuVisible}
            sidebarContent={<React.Suspense fallback={JavascriptMessage.loading.niceToString()}>
              <ToolbarRenderer
                sidebarExpanded={expandedSideMenu}
                closeSidebar={() => handleOpenSideBar(false)}
                sidebarFullScreen={fullSidebarView}
                logoExpanded={appIcon}
                appTitle={appName} />
            </React.Suspense>}
            sideMenuExpanded={expandedSideMenu}
            sideMenuHide={sideMenuHide}
            setSideMenuExpanded={handleOpenSideBar}
            fullSidebarView={fullSidebarView}>

            <VersionChangedAlert />

            <div className={"main-toopbar"} style={{ flexGrow: 0 }}>
              {AuthClient.currentUser() && sideMenuVisible && <a className="main-sidebar-button" onClick={(ev) => {
                window.dispatchEvent(new CustomEvent("sidebarMove"));
                if (!fullSidebarView) {
                  if (!expandedSideMenu && !sideMenuHide) {
                    setSideMenuHide(true);
                  } else {
                    setSideMenuHide(false);
                    setExpandedSideMenu(!expandedSideMenu);
                  }
                } else {
                  setExpandedSideMenu(!expandedSideMenu);
                }
              }}>
                <div style={{ display: "flex", height: "100%", alignItems: "center" }}>
                  <FontAwesomeIcon icon={"angle-double-left"} style={{ transition: "all 400ms", width: !expandedSideMenu && !sideMenuHide && !fullSidebarView ? "15px" : "0.1px" }} />
                  <FontAwesomeIcon icon={"bars"} style={{ transition: "all 400ms", width: (expandedSideMenu && !sideMenuHide) || fullSidebarView ? "15px" : "0.1px" }} />
                  <FontAwesomeIcon icon={"angle-double-right"} style={{ transition: "all 400ms", width: sideMenuHide && !fullSidebarView ? "15px" : "0.1px" }} />
                </div>
              </a>}

              {AuthClient.currentUser() && <div style={{ width: "100%", marginRight: "15px" }}>
                {AuthClient.isPermissionAuthorized(OmniboxPermission.ViewOmnibox) && <OmniboxAutocomplete inputAttrs={{ className: "form-control" }} />}
              </div>}

              {AuthClient.currentUser() && <React.Suspense fallback={null}><AlertDropdown /></React.Suspense>}

              {(!AuthClient.currentUser() || !sideMenuVisible) && <>
                {<div style={{ display: "flex", transition: "all 200ms", alignItems: "center", marginLeft: 10 }}>
                  <img className={"sidebar-brand-icon"} src={appIcon} />
                  <h5 className={"sidebar-app-title"} style={{ marginBottom: 0 }}>{appName}</h5>
                </div>}
                <div style={{ flexGrow: 1 }}></div>
                {!AuthClient.currentUser() && <CultureDropdown />}
              </>}

              <LoginDropdown changePasswordVisible={AuthClient.getAuthenticationType() != "azureAD"} />
            </div>

            <div id="page-inner-content" style={{ padding: 10, position: 'relative' }}>
              {Layout.switch}
            </div>
          </SidebarContainer>
        </div>
        <GlobalModalContainer />
        <div id="footer">
          <div className="container">
            <p className="text-muted">Made by <a href="http://signumsoftware.com/">Signum Software</a>  using <a href="http://signumframework.com/">Signum Framework</a>.</p>
          </div>
        </div>
      </div>
    </ErrorBoundary>
  );
}

Layout.switch = undefined as React.ReactElement<any> | undefined;
```

olmobrutall commented:
Huge change! Looks very nice already! 

Whenever I have some free time I will take a deeper look and try to apply the changes in southwind. 

Thanks a lot!

DCoder1997 commented:
I changed a big part of the sidebar and i describe the changes here:

- The sidebar, and the toolbar now have a interface called engine to configure the response of actions of the toggle and for refresh his items with across the new state.
-  I removed the location enum (Top, Middle, Bottom), and i disabled the union of the viewable toolbars. Reverted to priority system to determine the visible Toolbar, but for toolbar union, is possible to put a new toolbar element with other toolbar to make an union. The toolbar entity has a presaving to check the recursive calls of the toolbars to protect the database of a infinite loop.
- I changed in the layout the window listener
- I moved the toggle button of the sidebar to the SidebarContainer as new element with dependency of the engine interface, to make the code more clean and easy
- The props of the toolbar renderer are simplified with the engine interface, and i transformed the appIcon and appTitle to a React.Node to put your custom element.
- To resolve the workflow dependency and home icon i added possibility to put your custom children to the toolbar renderer, for example, graphs, custom images, home link, workflow dropdown and others.

The new example layout with everything working is like that:
```
export default function Layout() {
  const appName = "APPLICATION NAME";
  const appIcon = AppContext.toAbsoluteUrl("~/logo.svg");

  const itemStorageKeyExtend = "SIDEBAR_EXPANDS";
  const itemStorageKeyHide = "SIDEBAR_HIDE";

  const [refreshId, setRefreshId] = React.useState(0);
  const [sideMenuVisible] = React.useState(true);
  const [expandedSideMenu, setExpandedSideMenu] = React.useState(window.innerWidth > 991 && window.localStorage.getItem(itemStorageKeyExtend) == "1");
  const [sideMenuHide, setSideMenuHide] = React.useState(window.localStorage.getItem(itemStorageKeyHide) == "1");
  const [fullSidebarView, setFullSidebarView] = React.useState(window.innerWidth < 991);
  const sideMenuOpenRef = useUpdatedRef(expandedSideMenu);

  function handleOpenSideBar(to: boolean) {
    setExpandedSideMenu(to);
    setSideMenuHide(false);
  }

  useWindowEvent("resize", (ev) => {
    if (window.innerWidth <= 991) {
      if (!fullSidebarView) {
        setFullSidebarView(true);
        setExpandedSideMenu(false);
        setSideMenuHide(false);
      }
    }
    else {
      if (fullSidebarView) {
        setFullSidebarView(false);
        setExpandedSideMenu(true);
        setSideMenuHide(false);
      }
    }
  }, [fullSidebarView]);
   
  React.useEffect(() => {
    AppContext.Expander.onGetExpanded = () => !sideMenuOpenRef.current;
    AppContext.Expander.onSetExpanded = (isExpanded: boolean) => setExpandedSideMenu(!isExpanded);
  }, []); //Sidebar

  function resetUI() {
    setRefreshId(rID => rID + 1);
  };

  React.useEffect(() => {
    AppContext.setResetUI(resetUI);
    return () => AppContext.setResetUI(() => { });
  }, []);

  React.useEffect(() => {
    window.localStorage.setItem(itemStorageKeyExtend, expandedSideMenu ? "1" : "0");
    window.localStorage.setItem(itemStorageKeyHide, sideMenuHide ? "1" : "0");
  }, [sideMenuHide, expandedSideMenu]);

  var engine = {
    visible: AuthClient.currentUser() && sideMenuVisible,
    sideMenuExpanded: expandedSideMenu,
    setSideMenuExpanded: handleOpenSideBar,
    sideMenuHide: sideMenuHide,
    fullSidebarView: fullSidebarView,
    setSideMenuHide: setSideMenuHide,
    setExpandedSideMenu: setExpandedSideMenu,
  }

  return (
    <ErrorBoundary >
      <div id="main" key={refreshId}>
        <Notify />
        <div id="main-container">
          <SidebarContainer
            engine={engine}
            sidebarContent={<React.Suspense fallback={JavascriptMessage.loading.niceToString()}>
              <ToolbarRenderer
                engine={engine}
                logoExpanded={<img src={appIcon}></img>}
                appTitle={appName}>

                <Nav.Item>
                  <Nav.Link style={{ paddingLeft: expandedSideMenu === true ? "25px" : "13px" }}
                    onClick={(e: React.MouseEvent<any>) => { AppContext.pushOrOpenInTab(AppContext.toAbsoluteUrl("~/"), e); }}
                    onAuxClick={(e: React.MouseEvent<any>) => { AppContext.pushOrOpenInTab(AppContext.toAbsoluteUrl("~/"), e); }}
                    active={false}>
                    <FontAwesomeIcon icon={"home"} />
                    <span>{"Inicio"}</span>
                    {!expandedSideMenu && !fullSidebarView && <div className={"nav-item-float"}>{"Inicio"}</div>}
                  </Nav.Link>
                </Nav.Item>

                {/* Do you want workflow??
                 * <React.Suspense fallback={JavascriptMessage.loading.niceToString()}><WorkflowDropdown sidebarExpanded={p.sidebarExpanded} onClose={p.closeSidebar} fullScreenExpanded={p.sidebarFullScreen} onRefresh={() => { setTimeout(() => setRefresh(!refresh), 500); }} /></React.Suspense>*/}

              </ToolbarRenderer>
            </React.Suspense>}>

            <VersionChangedAlert />

            <div className={"main-toopbar"} style={{ flexGrow: 0 }}>
              <SidebarToggleItem engine={engine} />

              {AuthClient.currentUser() && <div style={{ width: "100%", marginRight: "15px" }}>
                {AuthClient.isPermissionAuthorized(OmniboxPermission.ViewOmnibox) && <OmniboxAutocomplete inputAttrs={{ className: "form-control" }} />}
              </div>}

              {AuthClient.currentUser() && <React.Suspense fallback={null}><AlertDropdown /></React.Suspense>}

              {(!AuthClient.currentUser() || !sideMenuVisible) && <>
                {<div style={{ display: "flex", transition: "all 200ms", alignItems: "center", marginLeft: 10 }}>
                  <img className={"sidebar-brand-icon"} src={appIcon} />
                  <h5 className={"sidebar-app-title"} style={{ marginBottom: 0 }}>{appName}</h5>
                </div>}
                <div style={{ flexGrow: 1 }}></div>
                {!AuthClient.currentUser() && <CultureDropdown />}
              </>}

              <LoginDropdown changePasswordVisible={AuthClient.getAuthenticationType() != "azureAD"} />
            </div>

            <div id="page-inner-content" style={{ padding: 10, position: 'relative' }}>
              {Layout.switch}
            </div>
          </SidebarContainer>
        </div>
        <GlobalModalContainer />
        <div id="footer">
          <div className="container">
            <p className="text-muted">Made by <a href="http://signumsoftware.com/">Signum Software</a>  using <a href="http://signumframework.com/">Signum Framework</a>.</p>
          </div>
        </div>
      </div>
    </ErrorBoundary>
  );
}

Layout.switch = undefined as React.ReactElement<any> | undefined;
```

doganc commented:
Perfect, do you publish any screen shot?

olmobrutall commented:
I have some but they are from their internal applications. I’ll make a proper notification when I have it integrated in southwind

DCoder1997 commented:
Yeah, i have it:

Before:
![image](https://user-images.githubusercontent.com/53967797/151414731-16201ea9-37d7-4403-993e-333586fd6a5c.png)

After::

Extended:
![image](https://user-images.githubusercontent.com/53967797/151415134-8208da7e-5780-4eb1-a3df-d6e566c2426b.png)

Mini:
![image](https://user-images.githubusercontent.com/53967797/151415374-621ea5f3-87d5-475e-990c-4e5c7a4da8b7.png)

Hide:
![image](https://user-images.githubusercontent.com/53967797/151415405-ca9de455-d9d9-427f-9491-c5fdd12ea1cb.png)

doganc commented:
I think, it's good, but i can't see The Omnibox, do you have it?

DCoder1997 commented:
The top bar is fully a omnibox, when you touich it the bar turns white.
![image](https://user-images.githubusercontent.com/53967797/151420699-6f54990a-e2f2-4f4d-b06e-040e2bacae5e.png)

like that
![image](https://user-images.githubusercontent.com/53967797/151420775-4848eaf4-e37a-435e-b60e-da29a796980d.png)

doganc commented:
Perfect

olmobrutall commented:
@DCoder1997  thanks a lot for this awesome sidebar!  

I've created `newToolbar` branches in Framework and Southwind and I've made some initial changes there.

This change is going to require manual adaptation work on every application so I want to be sure that the code in Layout.tsx is clean and flexible.

MehdyKarimpour commented:
Excellent job, DCoder1997 👌📣😀

On Fri, Jan 28, 2022 at 2:23 AM Olmo del Corral ***@***.***>
wrote:

> @DCoder1997 <https://github.com/DCoder1997> thanks a lot for this awesome
> sidebar!
>
> I've created newToolbar branches in Framework and Southwind and I've made
> some initial changes there.
>
> This change is going to require manual adaptation work on every
> application so I want to be sure that the code in Layout.tsx is clean and
> flexible.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/signumsoftware/framework/pull/536#issuecomment-1023714617>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTMTUWVAYSOQ737KL4CAJTUYHEFXANCNFSM5MX5BLRQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Your Sincerely
Mehdy Karimpour

[image: Green Eco]Please, consider the environment before printing this
email.

olmobrutall commented:
Current Status: 

I've working some hours in the new toolbar/layout. Some changes: 

* Use DirectedGraph fro Cycle checking, like Roles do. 
* Move CustomCultureDropdown to Translations module and make use of LoginDropdown.extraMenuITems.
* Remove SidebarEngine by converting many booleans to SidebarMode ("Wide" | "Narrow" | "Hidden")
* Add useBreakpoint / getBreakpoint hooks. 
* General cleanup of extra styles, css...
* Make the toolbar more navbar compatible, so it can be themed using https://bootswatch.com/ 
* Remove signum footer (at the end all the customers ask to remove it anyway...)
* improve server side infrastructure for CustomPermissionResponse
* improve client-side infrastructure for multiple ToolbarConfig per type

Also fix some responsiveness issues in SearchPage / SearchControl and #hero example. 

The only thing missing that I'm aware so far is cleaning the ToolbarWorkflowDropdown.

There are two posibilities, to resolve the Workflow.starts in the server or in the client. Both have his own issues :) 

All the changes are pushed in `newToolbar` branches in Framework and Southwind and will be merged soon.

DCoder1997 commented:
Thanks for your efforts. I know the possibilities to improve this, but due to lack of time I didn't have time to change everything to make it better. I tried to use DirectedGraph as a role system, but I don't know how to use it properly because I haven't used it long enough. To solve that, I made a simpler but fully working solution to check for recursion.

I wish to test it when is ready, i have some cases very sensitive to the code to test, like window size changes and check if is still works hahaha 😅😅

olmobrutall commented:
> Thanks for your efforts. I know the possibilities to improve this, but due to lack of time I didn't have time to change everything to make it better. I tried to use DirectedGraph as a role system, but I don't know how to use it properly because I haven't used it long enough. To solve that, I made a simpler but fully working solution to check for recursion.

Actually I haven't tried my version... The benefit of using DirectedGraph are that it also detects cycles when saving a ToolbarMenuEntity, and the error message says which connections are producing the cycle.

> I wish to test it when is ready, i have some cases very sensitive to the code to test, like window size changes and check if is still works hahaha 😅😅

I've tried responsiveness, I think it should work ok. I've changed the behaviour though: Now on desktop mode it flips from "Wide" <--> "Narrow" only, while in mobile mode if flips betheen "Wide" (with auto close) <--> "Hidden".

olmobrutall commented:
It's finally merged but there are still some open issue: 

* In Narrow mode the tooltip uses fixed, doesn't work with scrolling
* make ToolbarWorkflowDropdown work again

I will write a message tonight

MehdyKarimpour commented:
Perfect 👌

On Tue, Feb 1, 2022, 12:36 PM Olmo del Corral ***@***.***>
wrote:

> It's finally merged but there are still some open issue:
>
>    - In Narrow mode the tooltip uses fixed, doesn't work with scrolling
>    - make ToolbarWorkflowDropdown work again
>
> I will write a message tonight
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/signumsoftware/framework/pull/536#issuecomment-1026616504>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTMTUTQ3Y5EH24PHE2RL4TUY6PA3ANCNFSM5MX5BLRQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
